[2.7] Update task groups guide (#2653)
astronomer/docs: learn/task-groups.md

import task_group_example from '!!raw-loader!../code-samples/dags/task-groups/task_group_example.py';
import task_group_mapping_example from '!!raw-loader!../code-samples/dags/task-groups/task_group_mapping_example.py';

Airflow [task groups](https://airflow.apache.org/docs/apache-airflow/stable/core-concepts/dags.html#taskgroups) are a tool to organize tasks into groups within your DAGs. Using task groups allows you to:

- Organize complicated DAGs, visually grouping tasks that belong together in the Airflow UI **Grid View** and **Graph View**.
- Apply `default_args` to sets of tasks, instead of at the [DAG level](dags.md#dag-parameters).

> - [Dynamically map](dynamic-tasks.md) over task patterns creating dynamically mapped complex sets of tasks.

Reviewer: ```suggestion
- [Dynamically map](dynamic-tasks.md) over groups of tasks, enabling complex dynamic patterns.
```

- Turn task patterns into modules that can be reused across DAGs and even Airflow instances.

In this guide, you'll learn how to create and use task groups in your DAGs. You can find many example DAGs using task groups in the [webinar task group repository](https://github.com/astronomer/webinar-task-groups).

![Task group intro gif](/img/guides/task-groups_intro_task_group_gif.gif)

## Assumed knowledge

To get the most out of this guide, you should have an understanding of:

- Airflow operators. See [Operators 101](what-is-an-operator.md).

## When to use task groups

There are many use cases for task groups in Airflow DAGs and often they are used to visually organize complicated DAGs. For example, you might use task groups to organize tasks:

- In big ELT/ETL DAGs, where you could have a task group per table being modified, or per schema and leverage passing of different sets of `default_args` to each task group.
- In MLOps DAGs, where you could have a task group per model being trained. 

> - In Situations where several teams work on parts of the same DAG and you want to visually separate the tasks that belong to each team. In this case you might also want to consider separating the DAG into multiple DAGs and using [Datasets](airflow-datasets.md) to connect them.

Reviewer: ```suggestion
- In situations where several teams work on parts of the same DAG and you want to visually separate the tasks that belong to each team. In this case you might also want to consider separating the DAG into multiple DAGs and using [Datasets](airflow-datasets.md) to connect them.
```

- When you are using the same patterns of tasks in multiple DAGs and want to create a reusable module.

> Additionally when you have an input of unknown length, for example an unknown number of files in a directory, you can use task groups to [dynamically map](#generate-task-groups-dynamically-at-runtime) over the input and create a task group performing sets of actions for each file. This is the only way to dynamically map sequential tasks in Airflow.

Reviewer: Is there a reason this is separated out rather than being another bullet in the list?

Author: Hmm... good question. I think I separated it because in my mind it is an additional functional feature offering a DAG structure that is not possible without it. While all other examples are theoretically possible to build with tasks in DAGs without task groups, just improve the code a lot 😄 
I am open to making it a bullet point, maybe then with a slightly changed intro sentences, like "you can use task groups to:" instead of having "organize" in the intro

Reviewer: I think I would prefer adding it to the bullets and modifying the intro sentences.

## Define task groups

There are two ways to define task groups in your DAGs:

- Using the `TaskGroup` class to create a task group context
- Using the `@task_group` decorator on a Python function

> In most cases, it is a matter of personal preference in which way you like to define your task groups. The only exception is when you want to [dynamically map](dynamic-tasks.md) over a task group, this is only possible using `@task_group`. See [Airflow decorators](airflow-decorators.md) for more information on decorators and the TaskFlow API.

Reviewer: ```suggestion
In most cases, it is a matter of personal preference which method you use. The only exception is when you want to [dynamically map](dynamic-tasks.md) over a task group; this is only possible using `@task_group`. See [Airflow decorators](airflow-decorators.md) for more information on decorators and the TaskFlow API.
```

> The following code shows how to instantiate a simple task group using containing two sequential tasks using either method. You can use dependency operators (`<<` and `>>`) both in and with task groups in the same way that you can with individual tasks.

Reviewer: ```suggestion
The following code shows how to instantiate a simple task group containing two sequential tasks. You can use dependency operators (`<<` and `>>`) both within and between task groups in the same way that you can with individual tasks.
```

<Tabs
    defaultValue="decorator"
    groupId="define-task-groups"
    values={[
        {label: '@task_group', value: 'decorator'},
        {label: 'TaskGroup', value: 'context'},
    ]}>

<TabItem value="decorator">

```python
@task_group(
    group_id="task_group_1",
    default_args={"conn_id": "postgres_default"},
    tooltip="This task group is very important!",
    ui_color="#7352BA",  # Background color in graph view
    ui_fgcolor="#FFFFFF",  # Font color for in graph view
    prefix_group_id=True,
    # parent_group=None,
    # dag=None,
)
def tg1():
    t1 = EmptyOperator(task_id="t1")

tg1()
```

</TabItem>
<TabItem value="context">

```python
with TaskGroup(
    group_id="task_group_2",
    default_args={"conn_id": "postgres_default"},
    tooltip="This task group is also very important!",
    ui_color="#7352BA",  # Background color in graph view
    ui_fgcolor="#FFFFFF",  # Font color for in graph view
    prefix_group_id=True,
    # parent_group=None,
    # dag=None,
    # add_suffix_on_collision=True, # resolves group_id collisions by adding a suffix
) as tg2:
    t1 = EmptyOperator(task_id="t1")
```

</TabItem>
</Tabs>

## `task_id` in task groups

> When your task is within a task group, your callable `task_id` is the `task_id` prefixed with the `group_id`. For example, `group_id.task_id`. This ensures the task_id is unique across the DAG. It is important that you use this format when calling specific tasks with [XCom](airflow-passing-data-between-tasks.md) passing or [branching](airflow-branch-operator.md) operator decisions. You can disable this behavior by setting the [task group parameter](#task-group-parameters) `prefix_group_id=False`.

Reviewer: ```suggestion
When your task is within a task group, your callable `task_id` will be `group_id.task_id`. This ensures the task_id is unique across the DAG. It is important that you use this format when calling specific tasks when working with [XComs](airflow-passing-data-between-tasks.md) or [branching](airflow-branch-operator.md). You can disable this behavior by setting the [task group parameter](#task-group-parameters) `prefix_group_id=False`.
```

For example, the `task_1` task in the following DAG has a `task_id` of `my_outer_task_group.my_inner_task_group.task_1`.

<Tabs
    defaultValue="decorator"
    groupId="task_id-in-task-groups"
    values={[
        {label: '@task_group', value: 'decorator'},
        {label: 'TaskGroup', value: 'context'},
    ]}>

<TabItem value="decorator">

```python
@task_group(group_id="my_outer_task_group")
def my_outer_task_group():
    @task_group(group_id="my_inner_task_group")
    def my_inner_task_group():
        EmptyOperator(task_id="task_1")

    my_inner_task_group()

my_outer_task_group()
```

</TabItem>
<TabItem value="context">

```python
with TaskGroup(group_id="my_outer_task_group") as tg1:
    with TaskGroup(group_id="my_inner_task_group") as tg2:
        EmptyOperator(task_id="task_1")
```
</TabItem>
</Tabs>

## Passing data through task groups

> When using the `@task_group` decorator of the TaskFlow API you can pass data between before and after the task group as with regular `@task` decorators:

Reviewer: ```suggestion
When using the `@task_group` decorator of the TaskFlow API you can pass data through the task group as with regular `@task` decorators:
```

The following image shows the expanded view of the nested task groups in the Airflow UI:

![Nested task groups](/img/guides/task-groups_nested_tg.png)

## Custom task group classes

> If you find yourself using the same patterns of tasks in several DAGs or even several Airflow instances it may be useful to create a custom task group class module. To do so you need to inherit from the `TaskGroup` class and then define your tasks within that custom class. The task definitions will be the same as if you were defining them in a DAG file, with the only added requirement that you'll need to use `self` to assign the task to the task group.

Reviewer: ```suggestion
If you use the same patterns of tasks in several DAGs or Airflow instances, it may be useful to create a custom task group class module. To do so, you need to inherit from the `TaskGroup` class and then define your tasks within that custom class. The task definitions will be the same as if you were defining them in a DAG file, with the only added requirement that you'll need to use `self` to assign the task to the task group.
```
